README
NAME
    Benchmark::Serialize - Benchmarks of serialization modules

SYNOPSIS
        use Benchmark::Serialise qw(cmpthese);

        my $structure = {
            array  => [ 'a' .. 'j' ],
            hash   => { 'a' .. 'z' ]
            string => 'x' x 200,
        };

        cmpthese( -5, $structure, qw(:core :json :yaml) );

DESCRIPTION
    This module encapsulates some basic benchmarks to help you choose a
    module for serializing data. Note that using this module is only a part
    of chosing a serialization format. Other factors than the benchmarked
    might be of relevance!

    See the in-module documentation for more information.

RESULTS

  Modules
  Bencode             : 1.31
  Convert::Bencode    : 1.03
  Convert::Bencode_XS : 0.06
  Data::Dumper        : 2.125
  Data::Taxi          : 0.95
  FreezeThaw          : 0.45
  JSON::PP            : 2.26000
  JSON::XS            : 2.26
  PHP::Serialization  : 0.33
  RPC::XML            : 1.44
  Storable            : 2.21
  XML::Simple         : 2.18
  YAML::Old           : 0.81
  YAML::Tiny          : 1.40
  YAML::XS            : 0.32
  
  Sizes
                      bytes Bencode Convert::Bencode Convert::Bencode_XS YAML::XS JSON::XS JSON::PP YAML::Old YAML::Tiny Storable FreezeThaw XML::Simple PHP::Serialization Data::Taxi Data::Dumper RPC::XML
  Bencode               339      --               0%                  0%      -5%     -10%     -10%      -10%       -11%     -11%       -13%        -32%               -43%       -67%         -68%     -80%
  Convert::Bencode      339      0%               --                  0%      -5%     -10%     -10%      -10%       -11%     -11%       -13%        -32%               -43%       -67%         -68%     -80%
  Convert::Bencode_XS   339      0%               0%                  --      -5%     -10%     -10%      -10%       -11%     -11%       -13%        -32%               -43%       -67%         -68%     -80%
  YAML::XS              357      5%               5%                  5%       --      -5%      -5%       -5%        -6%      -7%        -9%        -29%               -40%       -65%         -66%     -79%
  JSON::XS              376     11%              11%                 11%       5%       --       0%       -0%        -1%      -2%        -4%        -25%               -37%       -63%         -64%     -78%
  JSON::PP              376     11%              11%                 11%       5%       0%       --       -0%        -1%      -2%        -4%        -25%               -37%       -63%         -64%     -78%
  YAML::Old             377     11%              11%                 11%       6%       0%       0%        --        -1%      -1%        -4%        -25%               -36%       -63%         -64%     -78%
  YAML::Tiny            379     12%              12%                 12%       6%       1%       1%        1%         --      -1%        -3%        -25%               -36%       -63%         -64%     -78%
  Storable              382     13%              13%                 13%       7%       2%       2%        1%         1%       --        -2%        -24%               -36%       -63%         -64%     -78%
  FreezeThaw            391     15%              15%                 15%      10%       4%       4%        4%         3%       2%         --        -22%               -34%       -62%         -63%     -77%
  XML::Simple           502     48%              48%                 48%      41%      34%      34%       33%        32%      31%        28%          --               -15%       -51%         -53%     -71%
  PHP::Serialization    593     75%              75%                 75%      66%      58%      58%       57%        56%      55%        52%         18%                 --       -42%         -44%     -65%
  Data::Taxi           1028    203%             203%                203%     188%     173%     173%      173%       171%     169%       163%        105%                73%         --          -3%     -40%
  Data::Dumper         1058    212%             212%                212%     196%     181%     181%      181%       179%     177%       171%        111%                78%         3%           --     -38%
  RPC::XML             1712    405%             405%                405%     380%     355%     355%      354%       352%     348%       338%        241%               189%        67%          62%       --
  
  Deflate
                          Rate YAML::Old RPC::XML Data::Taxi PHP::Serialization XML::Simple JSON::PP YAML::Tiny Bencode Convert::Bencode Data::Dumper FreezeThaw YAML::XS Storable Convert::Bencode_XS JSON::XS
  YAML::Old              241/s        --     -85%       -87%               -90%        -91%     -92%       -94%    -97%             -97%         -97%       -98%     -98%     -99%                -99%    -100%
  RPC::XML              1643/s      581%       --       -11%               -33%        -42%     -47%       -62%    -79%             -81%         -82%       -84%     -87%     -94%                -96%     -99%
  Data::Taxi            1845/s      665%      12%         --               -24%        -35%     -41%       -57%    -77%             -79%         -80%       -82%     -85%     -94%                -96%     -99%
  PHP::Serialization    2443/s      913%      49%        32%                 --        -13%     -22%       -43%    -69%             -72%         -74%       -76%     -81%     -91%                -94%     -99%
  XML::Simple           2823/s     1071%      72%        53%                16%          --      -9%       -34%    -64%             -68%         -70%       -72%     -78%     -90%                -93%     -98%
  JSON::PP              3116/s     1192%      90%        69%                28%         10%       --       -27%    -60%             -65%         -67%       -69%     -75%     -89%                -93%     -98%
  YAML::Tiny            4287/s     1678%     161%       132%                75%         52%      38%         --    -45%             -51%         -54%       -57%     -66%     -85%                -90%     -98%
  Bencode               7860/s     3160%     378%       326%               222%        178%     152%        83%      --             -11%         -16%       -22%     -37%     -73%                -82%     -96%
  Convert::Bencode      8805/s     3552%     436%       377%               260%        212%     183%       105%     12%               --          -5%       -12%     -30%     -69%                -79%     -95%
  Data::Dumper          9309/s     3761%     467%       404%               281%        230%     199%       117%     18%               6%           --        -7%     -26%     -68%                -78%     -95%
  FreezeThaw           10048/s     4067%     511%       445%               311%        256%     222%       134%     28%              14%           8%         --     -20%     -65%                -76%     -94%
  YAML::XS             12560/s     5109%     664%       581%               414%        345%     303%       193%     60%              43%          35%        25%       --     -56%                -71%     -93%
  Storable             28671/s    11791%    1645%      1454%              1074%        916%     820%       569%    265%             226%         208%       185%     128%       --                -33%     -84%
  Convert::Bencode_XS  42708/s    17612%    2499%      2215%              1648%       1413%    1271%       896%    443%             385%         359%       325%     240%      49%                  --     -76%
  JSON::XS            177750/s    73618%   10718%      9533%              7177%       6196%    5605%      4046%   2162%            1919%        1810%      1669%    1315%     520%                316%       --
  
  Inflate
                         Rate YAML::Old XML::Simple PHP::Serialization RPC::XML Data::Taxi JSON::PP YAML::Tiny Convert::Bencode FreezeThaw Bencode Data::Dumper YAML::XS Convert::Bencode_XS Storable JSON::XS
  YAML::Old             192/s        --        -45%               -51%     -60%       -78%     -83%       -89%             -91%       -93%    -95%         -98%     -99%               -100%    -100%    -100%
  XML::Simple           350/s       82%          --               -12%     -28%       -60%     -70%       -80%             -84%       -87%    -90%         -97%     -98%                -99%     -99%    -100%
  PHP::Serialization    395/s      106%         13%                 --     -18%       -54%     -66%       -78%             -82%       -85%    -89%         -96%     -97%                -99%     -99%    -100%
  RPC::XML              482/s      152%         38%                22%       --       -44%     -58%       -73%             -78%       -82%    -87%         -96%     -97%                -99%     -99%     -99%
  Data::Taxi            865/s      351%        147%               119%      79%         --     -25%       -51%             -61%       -67%    -76%         -92%     -94%                -98%     -99%     -99%
  JSON::PP             1152/s      501%        230%               191%     139%        33%       --       -35%             -48%       -56%    -68%         -89%     -92%                -98%     -98%     -99%
  YAML::Tiny           1777/s      827%        408%               350%     268%       105%      54%         --             -19%       -32%    -51%         -84%     -88%                -96%     -97%     -98%
  Convert::Bencode     2201/s     1048%        530%               457%     356%       154%      91%        24%               --       -16%    -39%         -80%     -85%                -96%     -97%     -97%
  FreezeThaw           2619/s     1266%        649%               563%     443%       203%     127%        47%              19%         --    -28%         -76%     -83%                -95%     -96%     -97%
  Bencode              3622/s     1789%        936%               816%     651%       319%     214%       104%              65%        38%      --         -67%     -76%                -93%     -94%     -96%
  Data::Dumper        10960/s     5616%       3035%              2673%    2173%      1167%     851%       517%             398%       318%    203%           --     -27%                -78%     -83%     -87%
  YAML::XS            15037/s     7742%       4202%              3704%    3018%      1639%    1205%       746%             583%       474%    315%          37%       --                -70%     -77%     -82%
  Convert::Bencode_XS 49321/s    25622%      14010%             12377%   10127%      5603%    4182%      2676%            2141%      1783%   1262%         350%     228%                  --     -25%     -41%
  Storable            65511/s    34066%      18641%             16473%   13485%      7475%    5587%      3587%            2877%      2401%   1709%         498%     336%                 33%       --     -22%
  JSON::XS            83510/s    43453%      23790%             21027%   17217%      9556%    7150%      4600%            3694%      3088%   2206%         662%     455%                 69%      27%       --
  
AUTHOR
    Peter Makholm, "<peter at makholm.net>"

BUGS
    Please report any bugs or feature requests to "bug-benchmark-serialize
    at rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Benchmark-Serialize>. I
    will be notified, and then you'll automatically be notified of progress
    on your bug as I make changes.

ACKNOWLEDGEMENTS
    This module started out as a script written by Christian Hansen, see
    http://idisk.mac.com/christian.hansen/Public/perl/serialize.pl

COPYRIGHT & LICENSE
    Copyright 2009 Peter Makholm.

    This program is free software; you can redistribute it and/or modify it
    under the terms of either: the GNU General Public License as published
    by the Free Software Foundation; or the Artistic License.

    See http://dev.perl.org/licenses/ for more information.



Changes
Changes for Benchmark::Serialize

0.07 - Thu Sep  9 09:24:53 2010
	Added benchmark for Data::MessagePack
	Improved handling of failing benchmarks
	Added option to test script to pass structure as perl code

0.06 - Wed Sep  8 16:01:03 2010
	Add test for idempotence of transformation

0.05 - Wed Sep  8 11:06:21 2010
	Added benchmark for Data::Pond

0.04 - Wed Jul 21 09:09:25 2010
    Added benchmark for JSON::DWIW and a prettyfied version of JSON::XS

0.03 - Tue Nov 24 13:56:42 2009
	Added the following benchmarks: PHP::Serilization::XS, Data::asXML, XML::TreePP
	Added :xml tag

0.02 - Mon Nov  2 09:58:15 2009
	Added comparison chart of sizes
	Export cmpthese (RT #50997)
	Fixed documentation errors (RT #50996)
	Fixed dependencies (RT #50998)

0.01 - Fri Oct 30 12:15:33 2009
	Refactored into a Perl Module

June 15, 2009
    - Available at http://gist.github.com/130005
    - Added Data::Dumper and a current set of YAML/JSON modules
    - Added Data::Dumper and a current set of YAML/JSON modules

Original
    - Available at
      http://idisk.mac.com/christian.hansen/Public/perl/serialize.pl


lib/Benchmark/Serialize.pm
package Benchmark::Serialize;

use strict;
use warnings;

=head1 NAME

Benchmark::Serialize - Benchmarks of serialization modules

=head1 VERSION

Version 0.07

=cut

our $VERSION = '0.07';

=head1 SYNOPSIS

    use Benchmark::Serialize qw(cmpthese);

    my $structure = {
        array  => [ 'a' .. 'j' ],
        hash   => { 'a' .. 'z' },
        string => 'x' x 200,
    };

    cmpthese( -5, $structure, qw(:core :json :yaml) );

=head1 DESCRIPTION

This module encapsulates some basic benchmarks to help you choose a module
for serializing data. Note that using this module is only a part of chosing a
serialization format. Other factors than the benchmarked might be of
relevance!

=head2 Functions

This module provides the following functions

=over 5

=item cmpthese(COUNT, STRUCTURE, BENCHMARKS ...)

Benchmark COUNT interations of a list of modules. A benchmark is either a name
of a supported module, a tag, or a hash ref containing at least an inflate, a
deflate, and a name attribute:

  {
      name    => 'JSON::XS',
      deflate => sub { JSON::XS::encode_json($_[0]) }
      inflate => inflate  => sub { JSON::XS::decode_json($_[0]) }
  }

By default Benchmark::Serialize will try to use the name attribute as a module
to be loaded. This can be overridden by having a packages attribute with an
arrayref containing modules to be loaded.

=back

=head2 Benchmark tags

The following tags are supported

=over 5

=item :all     - All modules with premade benchmarks 

=item :default - A default set of serialization modules

=item :core    - Serialization modules included in core

=item :json    - JSON modules

=item :yaml    - YAML modules

=item :xml     - XML formats

=back

=cut

use Benchmark          qw[timestr];
use Test::Deep::NoTest;
use UNIVERSAL::require qw[];

use Exporter qw(import);
our @EXPORT_OK   = qw( cmpthese );
our %EXPORT_TAGS = ( all => \@EXPORT_OK );

my $benchmarks = {
    'Bencode' => {
        deflate  => sub { Bencode::bencode($_[0])                },
        inflate  => sub { Bencode::bdecode($_[0])                }
    },    
    'Convert::Bencode' => {
        deflate  => sub { Convert::Bencode::bencode($_[0])       },
        inflate  => sub { Convert::Bencode::bdecode($_[0])       }
    },
    'Convert::Bencode_XS' => {
        deflate  => sub { Convert::Bencode_XS::bencode($_[0])    },
        inflate  => sub { Convert::Bencode_XS::bdecode($_[0])    }
    },
    'Data::asXML' => {
        deflate  => sub { Data::asXML->new(pretty=>0)->encode($_[0])->toString },
        inflate  => sub { Data::asXML->new(pretty=>0)->decode($_[0]) },
        xml      => 1,
    },
    'Data::Dumper' => {
        deflate  => sub { Data::Dumper->Dump([ $_[0] ])          },
        inflate  => sub { my $VAR1; eval $_[0]                   },
        default  => 1,
        core     => 1,
    },
    'Data::MessagePack' => {
        deflate  => sub { Data::MessagePack->pack($_[0])         },
        inflate  => sub { Data::MessagePack->unpack($_[0])       },
    },
    'Data::Taxi' => {
        deflate  => sub { Data::Taxi::freeze($_[0])              },
        inflate  => sub { Data::Taxi::thaw($_[0])                },
        xml      => 1,
    },
    'Data::Pond' => {
        deflate  => sub { Data::Pond::pond_write_datum($_[0])    },
        inflate  => sub { Data::Pond::pond_read_datum($_[0])     },
    },
    'Data::Pond,eval' => {
        deflate  => sub { Data::Pond::pond_write_datum($_[0])    },
        inflate  => sub { eval($_[0])                            },
        packages => ['Data::Pond'],
    },
    'FreezeThaw' => {
        deflate  => sub { FreezeThaw::freeze($_[0])              },
        inflate  => sub { FreezeThaw::thaw($_[0])                },
        default  => 1
    },
    'JSON::PP' => {
        deflate  => sub { JSON::PP::encode_json($_[0])           },
        inflate  => sub { JSON::PP::decode_json($_[0])           },
        default  => 1,
        json     => 1
    },
    'JSON::XS' => {
        deflate  => sub { JSON::XS::encode_json($_[0])           },
        inflate  => sub { JSON::XS::decode_json($_[0])           },
        default  => 1,
        json     => 1
    },
    'JSON::XS,pretty' => {
        deflate  => sub { $_[1]->encode( $_[0] ) },
        inflate  => sub { $_[1]->decode( $_[0] ) },
        args     => sub { JSON::XS->new->pretty(1)->allow_blessed(1)->convert_blessed(1)->canonical(1) },
        json     => 1,
        packages => ['JSON::XS'],
    },
    'JSON::DWIW' => {
        deflate  => sub { JSON::DWIW->to_json($_[0])             },
        inflate  => sub { JSON::DWIW::deserialize($_[0])         },
        json     => 1,
    },
    'Storable' => {
        deflate  => sub { Storable::nfreeze($_[0])               },
        inflate  => sub { Storable::thaw($_[0])                  },
        default  => 1,
        core     => 1,
    },
    'PHP::Serialization' => {
        deflate  => sub { PHP::Serialization::serialize($_[0])   },
        inflate  => sub { PHP::Serialization::unserialize($_[0]) }
    },
    'PHP::Serialization::XS' => {
        deflate  => sub { PHP::Serialization::XS::serialize($_[0])   },
        inflate  => sub { PHP::Serialization::XS::unserialize($_[0]) }
    },
    'RPC::XML' => {
        deflate  => sub { RPC::XML::response->new($_[0])->as_string         },
        inflate  => sub { RPC::XML::ParserFactory->new->parse($_[0])->value },
        packages => ['RPC::XML', 'RPC::XML::ParserFactory'],
        xml      => 1,
    },
    'YAML::Old' => {
        deflate  => sub { YAML::Old::Dump($_[0])                 },
        inflate  => sub { YAML::Old::Load($_[0])                 },
        default  => 1,
        yaml     => 1
    },
    'YAML::XS' => {
        deflate  => sub { YAML::XS::Dump($_[0])                  },
        inflate  => sub { YAML::XS::Load($_[0])                  },
        default  => 1,
        yaml     => 1
    },
    'YAML::Tiny' => {
        deflate  => sub { YAML::Tiny::Dump($_[0])                },
        inflate  => sub { YAML::Tiny::Load($_[0])                },
        default  => 1,
        yaml     => 1
    },
    'XML::Simple' => {
        deflate  => sub { XML::Simple::XMLout($_[0])             },
        inflate  => sub { XML::Simple::XMLin($_[0])              },
        default  => 1,
        xml      => 1,
    },
    'XML::TreePP' => {
        deflate => sub { XML::TreePP->new()->write( $_[0] )      },
        inflate => sub { XML::TreePP->new()->parse( $_[0] )      },
        xml     => 1,
    },
};

our $benchmark_deflate  = 1;       # boolean
our $benchmark_inflate  = 1;       # boolean
our $benchmark_size     = 1;       # boolean
our $output             = 'chart'; # chart or list

sub cmpthese {
    my $iterations = shift;
    my $structure  = shift;
    my %benchmark;
    for my $spec (@_) {
        if ( ref $spec eq "HASH" ) {
            $benchmark{ $spec->{name} } = $spec; 

        } elsif ( $spec eq "all" or $spec eq ":all" ) {
            $benchmark { $_ } = $benchmarks->{ $_ } for keys %{ $benchmarks };
        
        } elsif ( $spec eq "default" ) {
            $benchmark{ $_ } = $benchmarks->{ $_ } for grep { $benchmarks->{ $_ }->{default} } keys %{ $benchmarks };
        
        } elsif ( $spec =~ /^:(.*)/ ) {
            $benchmark{ $_ } = $benchmarks->{ $_ } for grep { $benchmarks->{ $_ }->{$1} } keys %{ $benchmarks };
        
        } elsif ( exists $benchmarks->{ $spec } ) {
            $benchmark{ $spec } = $benchmarks->{ $spec }
        
        } else {
            warn "Unknown benchmark '$spec'.";
        }
    }

    my $width   = width(keys %benchmark);
    my $results = { };

    print "\nModules\n";

    BENCHMARK:

    foreach my $name ( sort keys %benchmark ) {

        my $benchmark = $benchmark{$name};
        my @packages  = ( exists($benchmark->{packages}) ? @{ $benchmark->{packages} } : $name );
        
        $_->require or next BENCHMARK for @packages;

        $benchmark->{args} = [ $benchmark->{args}->() ] if exists $benchmark->{args}
                                                        && ref $benchmark->{args} eq "CODE";

        my ($deflated, $inflated);
        eval {
            $deflated = $benchmark->{deflate}->($structure, @{ $benchmark->{args} } );
            $inflated = $benchmark->{inflate}->($deflated,  @{ $benchmark->{args} } );
            1;
        } or do {
            warn "Benchmark $name died with:\n    $@\n";
            next BENCHMARK;
        };

        printf( "%-${width}s : %8s %s\n", $packages[0], $packages[0]->VERSION, 
                    eq_deeply($inflated, $structure) ? "Identity transformation" : "Changes representation" );

        $results->{deflate}->{$name} = timeit_deflate( $iterations, $structure, $benchmark )
            if $benchmark_deflate;

        $results->{inflate}->{$name} = timeit_inflate( $iterations, $structure, $benchmark )
            if $benchmark_inflate;

        $results->{size}->{$name}    = length( $deflated );
    }

    output( 'Sizes', "size", $output, $results->{size}, $width )
        if $benchmark_size;

    output( 'Deflate (perl -> serialized)', "time", $output, $results->{deflate}, $width )
        if $benchmark_deflate;

    output( 'Inflate (serialized -> perl)', "time", $output, $results->{inflate}, $width )
        if $benchmark_inflate;
}

sub output {
    my $title  = shift;
    my $type   = shift;
    my $output = shift;
    printf( "\n%s\n", $title );
    if ( $type eq "size" ) {
        ($output eq "list") ? &size_list : &size_chart ; 
    } elsif ( $type eq "time" ) {
        ($output eq "list") ? &time_list : &time_chart ; 

    } else {
        warn("Unknown data type: $type");
    }
}

sub time_chart {
    my $results = shift;
    Benchmark::cmpthese($results);
}

sub time_list {
    my $results = shift;
    my $width   = shift;
    foreach my $title ( sort keys %{ $results } ) {
        printf( "%-${width}s %s\n", $title, timestr( $results->{ $title } ) );
    }
}

sub size_chart {
    my $results = shift;
    my @vals    = sort { $a->[1] <=> $b->[1] } map { [ $_, $results->{$_} ] } keys %$results;

    my @rows    = ( [
        '',
        'bytes',
        map { $_->[0] } @vals,
    ] );

    my @col_width = map { length ( $_ ) } @{ $rows[0] };

    for my $row_val ( @vals ) {
        my @row;

        push @row, $row_val->[0], $row_val->[1];
        $col_width[0] = ( length ( $row_val->[0] ) > $col_width[0] ? length( $row_val->[0] ) : $col_width[0] );
        $col_width[1] = ( length ( $row_val->[1] ) > $col_width[1] ? length( $row_val->[1] ) : $col_width[1] );

        # Columns 2..N = performance ratios
        for my $col_num ( 0 .. $#vals ) {
            my $col_val = $vals[$col_num];
            my $out;

            if ( $col_val->[0] eq $row_val->[0] ) {
                $out = "--";
            } else {
                $out = sprintf( "%.0f%%", 100*$row_val->[1]/$col_val->[1] - 100 );
            }

            push @row, $out;
            $col_width[$col_num+2] = ( length ( $out ) > $col_width[$col_num+2] ? length ( $out ) : $col_width[$col_num+2]);
        }
        push @rows, \@row;
    }

    # Pasted from Benchmark.pm
    # Equalize column widths in the chart as much as possible without
    # exceeding 80 characters.  This does not use or affect cols 0 or 1.
    my @sorted_width_refs = 
       sort { $$a <=> $$b } map { \$_ } @col_width[2..$#col_width];
    my $max_width = ${$sorted_width_refs[-1]};

    my $total = @col_width - 1 ;
    for ( @col_width ) { $total += $_ }

    STRETCHER:
    while ( $total < 80 ) {
        my $min_width = ${$sorted_width_refs[0]};
        last
           if $min_width == $max_width;
        for ( @sorted_width_refs ) {
            last 
                if $$_ > $min_width;
            ++$$_;
            ++$total;
            last STRETCHER
                if $total >= 80;
        }
    }

    # Dump the output
    my $format = join( ' ', map { "%${_}s" } @col_width ) . "\n";
    substr( $format, 1, 0 ) = '-';
    for ( @rows ) {
        printf $format, @$_;
    }
}

sub size_list {
    my $results = shift;
    my $width   = shift;
    foreach my $title ( sort keys %{ $results } ) {
        printf( "%-${width}s : %d bytes\n", $title, $results->{ $title } );
    }
}

sub timeit_deflate {
    my ( $iterations, $structure, $benchmark ) = @_;
    my $deflate = $benchmark->{deflate};
    return Benchmark::timethis( $iterations, sub { &$deflate($structure, @{ $benchmark->{args}}) }, '', 'none' );
}

sub timeit_inflate {
    my ( $iterations, $structure, $benchmark ) = @_;
    my $inflate = $benchmark->{inflate};
    my $deflated = $benchmark->{deflate}->($structure, @{ $benchmark->{args} });
    return Benchmark::timethis( $iterations, sub { &$inflate($deflated, @{ $benchmark->{args} }) }, '', 'none' );
}

sub width {
    return length( ( sort { length $a <=> length $b } @_ )[-1] );
}

=head1 RESULTS

See the README file for example results.

=head1 AUTHOR

Peter Makholm, C<< <peter at makholm.net> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-benchmark-serialize at
rt.cpan.org>, or through
the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Benchmark-Serialize>.  I will
be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 ACKNOWLEDGEMENTS

This module started out as a script written by Christian Hansen, see 
http://idisk.mac.com/christian.hansen/Public/perl/serialize.pl

=head1 COPYRIGHT & LICENSE

Copyright 2009 Peter Makholm.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


=cut

1;
